### PR TITLE
Add Hatch

### PR DESCRIPTION
**Pull request by thomaspinder**
## Type of changes

- [ ] Bug fix
- [X] New feature
- [ ] Documentation / docstrings
- [ ] Tests
- [ ] Other

## Checklist

- [X] I've formatted the new code by running `poetry run pre-commit run --all-files --show-diff-on-failure` before committing.
- [X] I've added tests for new code.
- [X] I've added docstrings for the new code.

## Description

This PR migrates the package manager from Poetry to Hatch. This is inline with Python's support and forward-looking guidance. Additionally, scripts to format source code and examples is now added. After running these formatters, several notebooks have aesthetically changed. Finally, `uv` is now set to be the default package installer. This results in faster build times.

Issue Number: N/A

### CODE REVIEW

File: docs/contributing.md [hunk 1]
```diff
@@ -72,36 +72,19 @@ you through every detail!
...
   ```
 
   At this point we recommend you check your installation passes the supplied unit tests:
 
   ```bash
-  $ poetry run pytest
+  $ hatch run dev:all-tests
   ```
 
-5.  Install the pre-commit hooks.
```
Comment (frazane) [inline docs/contributing.md:88]: No pre-commit hooks with Hatch?

File: pyproject.toml [hunk 2]
```diff
@@ -1,85 +1,146 @@
...
-]
-license = "Apache-2.0"
+dynamic = ["version"]
+description = 'Gaussian processes in JAX.'
 readme = "README.md"
-homepage = "https://github.com/JaxGaussianProcesses/GPJax"
-repository = "https://github.com/JaxGaussianProcesses/GPJax"
-include = ["LICENSE"]
-documentation = "https://docs.jaxgaussianprocesses.com/"
+requires-python = ">=3.9, <3.13"
```
Comment (frazane) [inline pyproject.toml:10]: I think it's supposed to be `>=3.10`

File: pyproject.toml [hunk 3]
```diff
@@ -1,85 +1,146 @@
...
+  "black",
+  "jupytext",
+]
+
+[tool.hatch.envs.dev.scripts]
+black-format = ["black ./gpjax ./tests", "jupytext --pipe black examples/*.py"]
+black-check = "black --check --diff ./gpjax ./tests"
+imports-format = [
+  "isort ./gpjax ./tests",
+  'isort examples/*.py --treat-comment-as-code "# %%" --float-to-top',
```
Comment (frazane) [inline pyproject.toml:97]: Nit - single vs double quotes are not consistent in this file

File: docs/contributing.md (hunk 1 above)
Comment (thomaspinder) [reply docs/contributing.md:88]: No. 1) A lot of people weren't using them anyway, and 2) I haven't figured out how to make them play nice with Hatch's environments. If it becomes a problem, then we can always reintroduce the pre-commit.

**Review by thomaspinder — COMMENTED** [review]
Respond to PR comments.

**Review by frazane — APPROVED** [review]

**Pull request merged**